src/lib/services/winners-service.ts

```diff
@@ -3,23 +3,69 @@ import { PockyDB } from '../database/db-interfaces';
 import TableHelper from '../parsers/tableHelper';
 import __logger from '../logger';
 import { Receiver } from '../../models/receiver';
+import Config from '../config';
 
 export interface WinnersService {
+	getWinners(results: ResultRow[]): ResultRow[]
 	returnWinnersResponse(): Promise<string>
 }
 
 export class DefaultWinnersService implements WinnersService {
 	readonly cannotDisplayResults: string = 'Error encountered; cannot display winners.';
 	database : PockyDB;
+	config : Config;
 
-	constructor(database : PockyDB) {
+	constructor(database : PockyDB, config: Config) {
 		this.database = database;
+		this.config = config;
+	}
+
+	getWinners(results: ResultRow[]): ResultRow[] {
+		let goodSenders : { senderid : string, amtReceived : number, received: ResultRow[] }[] = []
+		const keywords = this.config.getStringConfig('keyword');
+		const penaltyKeywords = this.config.getStringConfig('penaltyKeyword');
+		const minimum = this.config.getConfig('minimum');
+		const requireKeywords = this.config.getConfig('requireValues');
+
+		const allSenders = results.map(x => x.senderid).sort()
+			.filter((value, index, array) => index === 0 || value !== array[index - 1]);
+
+		const pegValid = (comment: string) => {
+			if (requireKeywords) {
+				return keywords.some(x => comment.toLowerCase().includes(x.toLowerCase()));
+			} else {
+				return !penaltyKeywords.some(x => comment.toLowerCase().includes(x.toLowerCase()));
+			}
+		}
+
+		allSenders.forEach(sender => {
+			const goodPegs = results.filter(x => x.senderid === sender && pegValid(x.comment));
+
+			if (goodPegs.length >= minimum) {
+				const received = results.filter(x => x.receiverid === sender && pegValid(x.comment));
+				goodSenders.push({
+					senderid : sender,
+					amtReceived : received.length,
+					received
+				});
+			}
+		});
+
+		let topReceived = goodSenders.map(x => x.amtReceived).sort().reverse()
+			.filter((value, index, array) => index === 0 || value !== array[index - 1])
+			.slice(0, this.config.getConfig('winners'));
```

Comment:
Won't this give 3 winners every time, regardless of ties?

Reply:
Also, what if there are less than the winners config of goodsenders? Won't that give an index out of bounds exception?

Reply:
Nope, this filters out the top three winning received numbers, which then gets used in the next bit to get the winners

Reply:
Also nope, slice doesn't care about the actual length of the array, and will just truncate to however many items it actually has to put in it. This is why the unit tests don't fail.